### ISSUE TITLE
version constraints are somewhat not respected and break concretizer

### ISSUE THREAD

**Issue by davydden** [spack/spack#5753]
with https://github.com/LLNL/spack/pull/5752 `slepc` will correctly have
```
depends_on('petsc@3.7:3.7.7', when='@3.7.1:3.7.4')
```
so I expected the current stable release of `slepc@3.7.4` to result in `petsc@3.7.7`, yet this does not happen:

```
$ spack spec -I dealii
Input spec
--------------------------------
     dealii

Normalized
--------------------------------
     dealii
         ^blas
         ^bzip2
         ^cmake
         ^lapack
         ^muparser
         ^suite-sparse
         ^tbb
         ^zlib

Concretized
--------------------------------
==> Error: An unsatisfiable version constraint has been detected for spec:

    petsc@3.8.0%clang@9.0.0-apple+boost~complex~debug+double+hdf5+hypre~int64+metis+mpi+mumps+shared+superlu-dist~trilinos arch=darwin-highsierra-x86_64


while trying to concretize the partial spec:

    slepc@3.7.4%clang@9.0.0-apple+arpack arch=darwin-highsierra-x86_64
        ^python@2.7.14%clang@9.0.0-apple+pic~shared~tk~ucs4 arch=darwin-highsierra-x86_64
            ^bzip2@1.0.6%clang@9.0.0-apple+shared arch=darwin-highsierra-x86_64
            ^ncurses@6.0%clang@9.0.0-apple~symlinks arch=darwin-highsierra-x86_64
                ^pkg-config@0.29.2%clang@9.0.0-apple+internal_glib arch=darwin-highsierra-x86_64
            ^openssl@1.0.2k%clang@9.0.0-apple arch=darwin-highsierra-x86_64
                ^zlib@1.2.11%clang@9.0.0-apple+pic+shared arch=darwin-highsierra-x86_64
            ^readline@7.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64
            ^sqlite@3.20.0%clang@9.0.0-apple arch=darwin-highsierra-x86_64


slepc requires petsc version 3.7:3.7.7, but spec asked for 3.8.0
```

So it looks like `petsc` got settled with `3.8.0` before trying to concretize `slepc`. 
Not sure if it's a bug or limitation of the current concretizer. 

@tgamblin @scheibelp ping

p.s. a workaround is to add to `~/.spack/packages.yaml`:
```
  petsc:
     version: [3.7.7]
```

For completeness, here are my settings:
```
packages:
  all:
    compiler: [clang, gcc]
    providers:
      mpi: [openmpi]
      blas: [openblas]
      lapack: [openblas]
      scalapack: [netlib-scalapack]
  cmake:
    version: [3.9.0]
  suite-sparse:
    variants: ~tbb
  boost:
    variants: +thread+system+serialization+iostreams+mpi+python
  petsc:
    variants: +mumps
  dealii:
    version: [develop]
    variants: +doc+optflags+adol-c+nanoflann+sundials+assimp
  metis:
    variants: +real64
```

**Comment by davydden** [comment]
there are four packages which use `slepc`:
```
$ ag "depends_on.*slepc" var/spack/repos/builtin/packages/
var/spack/repos/builtin/packages/py-slepc4py/package.py
38:    depends_on('slepc')

var/spack/repos/builtin/packages/fenics/package.py
85:    depends_on('slepc@3.4:', when='+slepc')

var/spack/repos/builtin/packages/gmsh/package.py
72:    depends_on('slepc', when='+slepc+petsc')

var/spack/repos/builtin/packages/dealii/package.py
146:    depends_on("slepc",            when='+slepc+petsc+mpi')
147:    depends_on("slepc@:3.6.3",     when='@:8.4.1+slepc+petsc+mpi')
148:    depends_on("slepc~arpack",     when='+slepc+petsc+mpi+int64')
```
and **all of them** fail with "unsatisfiable version constraint has been detected for spec" ;-(

**Comment by scheibelp** [comment]
> So it looks like petsc got settled with 3.8.0 before trying to concretize slepc.
Not sure if it's a bug or limitation of the current concretizer.

This is a limitation of the current concretizer: see https://github.com/LLNL/spack/issues/5465 and more-specifically https://github.com/LLNL/spack/issues/5465#issuecomment-332027738 for more info.

**Comment by davydden** [comment]
@scheibelp 

> see #5465 and more-specifically #5465 (comment) for more info.

I am not sure it's the same issue, as there are no packages in the graph depending on `petsc` other than the root `dealii` and `slepc`.

**Comment by alalazo** [comment]
Duplicate of #1781

**Comment by alalazo** [comment]
The issue is the same as it was for #1781. A node in the DAG has an unconditional dependency on `slepc` and the latest version of `slepc` is selected by the original concretizer. Later in the iterative process we meet a constraint on that version and fail since the original concretizer doesn't backtrack.

**TL;DR** Will be solved by #19501 if using the ASP based concretizer.
